compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6459,10 +6459,15 @@ private List<BLangNamedArgsExpression> checkProvidedErrorDetails(BLangErrorConst
         for (BLangNamedArgsExpression namedArgsExpression : errorConstructorExpr.namedArgs) {
             BType target = checkErrCtrTargetTypeAndSetSymbol(namedArgsExpression, expectedType);
 
+            namedArgsExpression.cloneAttempt++;
             BLangNamedArgsExpression clone = nodeCloner.cloneNode(namedArgsExpression);
-            BType type = checkExpr(clone, target, data);
+            BType type = checkExprSilent(clone, target, data);
             if (type == symTable.semanticError) {
-                checkExpr(namedArgsExpression, data);
+                if (Types.getReferredType(target).tag == TypeTags.RECORD) {
```

Comment:
Why special case just the record type? Shouldn't that be any non union?

Reply:
```type MyError2 error <map<string>>;
function testErrorTypeRefNegative() {
    MyError2 me2 = error MyError2("Message", c = 2); // invalid arg type in error detail field 'c', expected "
                                                     //                   "'string', found 'int'
}``` 
if we consider c=2, the error should be as mentioned in the comment.  
To get the type of c as int, in the checkExpr target should noType 
then we will evaluate the type as int. This is one of the non union case 
has to be go through else statement. But only for the record type, we have to check 
the availability of individual fields.

Reply:
+1 to type check with `target` for non-unions.

Also, we will need to modify the below to check if the `field.type` is union and is not assignable.  (to avoid duplicate diagnostics as we already type checked for non unions and logged errors)
https://github.com/ballerina-platform/ballerina-lang/blob/3d4e9517b98d902383d0d49282e29736d2804d78/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L3349-L3351